File: Difficulty/B. Medium/0430-flatten-a-multilevel-doubly-linked-list/0430-flatten-a-multilevel-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
public:
    /**
     * @brief Flattens a multilevel doubly linked list into a single-level doubly linked list.
     * 
     * @param head (Node*): The head of the multilevel doubly linked list.
     * 
     * @return Node*: The head of the flattened doubly linked list.
     */
    Node* flatten(Node* head) {
        Node* currentNode = head;  // Pointer to traverse the main list

        // Traverse through the entire list
        while (currentNode != nullptr) {
            // If the current node has a child, we need to flatten the child list
            if (currentNode->child != nullptr) {
                // Merge the child list with the main list
                mergeChildList(currentNode);
            }

            // Move to the next node in the list
            currentNode = currentNode->next;
        }

        // Return the modified (flattened) list
        return head;
    }

private:
    /**
     * @brief Merges the child list of the current node into the main doubly linked list.
     * 
     * @param currentNode (Node*): The node whose child list needs to be merged.
     */
    void mergeChildList(Node* currentNode) {
        Node* childListHead = currentNode->child;  // Head of the child list
        Node* childListTail = childListHead;  // We will find the tail of the child list

        // Traverse the child list to find the tail node
        while (childListTail->next != nullptr) {
            childListTail = childListTail->next;
        }

        // If the current node has a next node, connect it to the tail of the child list
        if (currentNode->next != nullptr) {
            childListTail->next = currentNode->next;  // Connect tail of the child list to the next node
            currentNode->next->prev = childListTail;  // Set the prev pointer of the next node
        }

        // Connect the current node to the head of the child list
        currentNode->next = childListHead;
        childListHead->prev = currentNode;

        // Remove the child pointer since the child list has been flattened
        currentNode->child = nullptr;
    }
};

File: Explore Card Track/3. Linked List/0430-flatten-a-multilevel-doubly-linked-list/0430-flatten-a-multilevel-doubly-linked-list.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, prev, next, child):
        self.val = val
        self.prev = prev
        self.next = next
        self.child = child
"""

class Solution:
    def flatten(self, head: 'Optional[Node]') -> 'Optional[Node]':
        """
        Flattens a multilevel doubly linked list into a single-level doubly linked list.

        Parameters:
        head (Optional[Node]): The head of the multilevel doubly linked list.

        Returns:
        Optional[Node]: The head of the flattened doubly linked list.
        """
        current_node = head  # Pointer to the current node we are traversing
        
        while current_node is not None:
            # If the current node has a child, we need to flatten the child list into the main list
            if current_node.child is not None:
                # Merge the child list starting from the current node
                self._merge_child_list(current_node)
            
            # Move to the next node in the parent list
            current_node = current_node.next
        
        return head

    def _merge_child_list(self, current_node: 'Node') -> None:
        """
        Merges the child list into the main doubly linked list at the current node.

        Parameters:
        current_node (Node): The node in the parent list that has a child list to be merged.
        """
        child_list_head = current_node.child  # The start of the child list
        
        # Traverse to the end of the child list to get the last node
        child_list_tail = child_list_head
        while child_list_tail.next is not None:
            child_list_tail = child_list_tail.next
        
        # If the current node has a next node, connect the tail of the child list to it
        if current_node.next is not None:
            child_list_tail.next = current_node.next  # Connect the child list tail to the current next node
            current_node.next.prev = child_list_tail  # Update the prev pointer of the next node
        
        # Now connect the current node to the child list's head
        current_node.next = current_node.child  # Link current node to the start of the child list
        current_node.child.prev = current_node  # Update the prev pointer of the child list head

        # Remove the child pointer as the child list has been flattened
        current_node.child = None
        